charon/src/bin/charon-driver/translate/translate_bodies.rs

```diff
@@ -566,28 +566,42 @@ impl BodyTransCtx<'_, '_, '_> {
                                     self.translate_constant_expr_to_const_generic(span, len)?;
                                 UnsizingMetadata::Length(len)
                             }
-                            hax::UnsizingMetadata::VTablePtr(impl_expr) => {
+                            hax::UnsizingMetadata::DirectVTable(impl_expr) => {
                                 let tref = self.translate_trait_impl_expr(span, impl_expr)?;
-                                match &impl_expr.r#impl {
+                                let vtable = match &impl_expr.r#impl {
                                     hax::ImplExprAtom::Concrete(tref) => {
                                         // Ensure the vtable type is translated.
-                                        let _: GlobalDeclId = self.register_item(
+                                        let id = self.register_item(
                                             span,
                                             tref,
                                             TransItemSourceKind::VTableInstance(
                                                 TraitImplSource::Normal,
                                             ),
                                         );
+                                        Some(GlobalDeclRef {
+                                            // FIXME: this is wrong, I believe
+                                            generics: Box::new(GenericArgs::empty()),
+                                            id,
+                                        })
                                     }
                                     // TODO(dyn): more ways of registering vtable instance?
                                     _ => {
                                         trace!(
                                             "impl_expr not triggering registering vtable: {:?}",
                                             impl_expr
-                                        )
+                                        );
+                                        None
```

Comment:
You'll notice that even if we return `ConstantExprKind::VTable` in this branch, we can't register the vtable instance since we don't know the impl. That's a fundamental incompleteness in the choice I made to translate vtable instances lazily: it won't work for:
```rust
fn foo<T: Trait>(x: &T) -> &dyn Trait { x }
fn main() {
    let _ = foo::<SomeType>(...);
}
```
You might want a `--always-translate-vtable-instances` for your purposes. Or maybe we should do a static analysis at the end of translation to register all the ones that might be used indirectly like that :thinking:. Or maybe if we see any `dyn Trait` anywhere then we just register all the vtable instances for that trait.